### PR TITLE
PCSX2-GUI: Rehaul BIOS / Folder Selector

### PR DESCRIPTION
**Pull request by RedDevilus**
### Description of Changes

- Remove button BIOS / Folder Selector in GUI
- Move BIOS / Folder Selector contents in General Settings
### Rationale behind Changes

- Since emulation settings has been renamed general settings, we can put BIOS and Folder in the same location where people have an overview and can just use tab key on icons or mouse. Less confusion and also fixes the speedhacks section *for the zillionth time*.
### Suggested Testing Steps

- Test if runtime wizard works, game starts and esspecialy emulation settings (while checking size of the window is good enough)

### CODE REVIEW

**Comment by RedDevilus** [comment]
On my PC:

image
FirstTimeSetup
image

**Review by kleinkkk — APPROVED** [review]

**Comment by kozarovv** [comment]
I'm not sure that is good idea to put two least used options on top of column.

**Comment by RedDevilus** [comment]
Yeah and it's a mess to work correctly. But im bot against putting at bottom which is what i originally did.

**Comment by Shoegzer** [comment]
Nice idea @RedDevilus.  Just a thought, but perhaps it's worth considering not having a separate section just for BIOS.  In other words, move the "BIOS Search Path" option to the bottom of the folders section and simply require the user to select a BIOS ROM file vs. a path.  This would at least keep all "folder/file selection" options in one place while retaining consistency with most other emulators in the way BIOS files are chosen.  The "Select a BIOS ROM" window could be retained if it was absolutely necessary though I wouldn't miss it.  Thoughts?

**Comment by RedDevilus** [comment]
I'd rather have something like what DobieStation has:
Give a set path for browse+ pick bios.

I constantly have to browse to my correct path whenever i get development builds. I've noticed there are bigger issues lying beneath the BIOS / folder selector which makes it very hard to me and would need some help with the multi inheritance situation. I did increase the height as temporary fix so it could technically fit but WX just scares me and it will be torn out when QT comes so i have some thinking to do.

**Comment by Shoegzer** [comment]
Fair points, they make sense to me.  And moving the BIOS selector to the folders area could be just food for thought until the Qt move anyway, I don't blame you at all for being WX-gunshy. ;)

File: 3rdparty/wxwidgets3.0/src/common/sizer.cpp [hunk 1]
```diff
@@ -1040,10 +1040,10 @@ wxSize wxSizer::GetMinSize()
     return ret;
 }
 
-void wxSizer::DoSetMinSize( int width, int height )
+void wxSizer::DoSetMinSize(int width, int height)
 {
-    m_minSize.x = width;
-    m_minSize.y = height;
+	m_minSize.x = width;
+	m_minSize.y = height * 1.6; // Hackfix: There is no real fixed height, workaround for general settings where BIOS / Folders / GS Window / Speedhacks and others reside in.
```
Comment (GovanifY) [inline 3rdparty/wxwidgets3.0/src/common/sizer.cpp:1046]: no no no NO. no 3rdparty patches please damnit

**Review by GovanifY — CHANGES_REQUESTED** [review]
LGTM code wise, others will have to approve on the design though; I don't mind either way.

Except for that 3rdparty patch. What even is this? Why do you add a hack on wxwidgets??

File: 3rdparty/wxwidgets3.0/src/common/sizer.cpp (hunk 1 above)
Comment (RedDevilus) [reply 3rdparty/wxwidgets3.0/src/common/sizer.cpp:1046]: Look at this problem: https://github.com/PCSX2/pcsx2/pull/4580

Comment (RedDevilus) [reply 3rdparty/wxwidgets3.0/src/common/sizer.cpp:1046]: Buttons like ok and cancel are dpi sensitive and restore restore defaults just dissapears. It's also bad on other languag

Comment (GovanifY) [reply 3rdparty/wxwidgets3.0/src/common/sizer.cpp:1046]: see how we handle multi-DPI scaling here: https://github.com/PCSX2/pcsx2/blob/master/pcsx2/gui/Dialogs/AboutBoxDialog.cpp#L38

also either way the bug isn't the issue here, the solution you're proposing is the issue! there is approximately 0% of chance seeing your pr merged with a 3rdparty hack

Comment (RedDevilus) [reply 3rdparty/wxwidgets3.0/src/common/sizer.cpp:1046]: I'll fix it then.

**Comment by vsub** [comment]
That's nice but it makes the window longer
How about place that side panel at the top(my poor photoshop attempt)and maybe add the memory cards too so less dialogs
Untitled-1

**Comment by lightningterror** [comment]
I believe it used to be at the top in the past but had issues?

**Review by kleinkkk — APPROVED** [review]

**Comment by lightningterror** [comment]
Status?

**Comment by RedDevilus** [comment]
Still need to fix the height factor in a non-hacky way.

**Comment by lightningterror** [comment]
> Still need to fix the height factor in a non-hacky way.

Hmm, if the issue is also present on master we can still merge the current changes anyway.

**Comment by refractionpcsx2** [comment]
> > Still need to fix the height factor in a non-hacky way.
> 
> Hmm, if the issue is also present on master we can still merge the current changes anyway.

The hack requires modifying WX, and we don't really want to do that.

**Comment by lightningterror** [comment]
> > > Still need to fix the height factor in a non-hacky way.
> > 
> > 
> > Hmm, if the issue is also present on master we can still merge the current changes anyway.
> 
> The hack requires modifying WX, and we don't really want to do that.

I meant excluding the hackfix, which can be fixed in a different pr later on if it's an issue on master as well.

**Comment by RedDevilus** [comment]
I suppose, WX is smart enough to add sliders for the icons so it's a minor inconvenience for now. I'll remove the hackfix.

**Comment by RedDevilus** [comment]
I have removed the hackfix.

Hackfix:
image

Current:
image

Keep in mind there is some differences depending on resolution and DPI.

File: 3rdparty/wxwidgets3.0/src/common/sizer.cpp [hunk 2]
```diff
@@ -1040,10 +1040,10 @@ wxSize wxSizer::GetMinSize()
     return ret;
 }
 
-void wxSizer::DoSetMinSize( int width, int height )
+void wxSizer::DoSetMinSize(int width, int height)
 {
-    m_minSize.x = width;
-    m_minSize.y = height;
+	m_minSize.x = width;
+	m_minSize.y = height;
```
Comment (lightningterror) [inline 3rdparty/wxwidgets3.0/src/common/sizer.cpp:1046]: Need to undo formatting changes too.

File: pcsx2/gui/Dialogs/FirstTimeWizard.cpp [hunk 3]
```diff
@@ -228,5 +228,5 @@ void FirstTimeWizard::OnPageChanging( wxWizardEvent& evt )
 void FirstTimeWizard::OnPageChanged( wxWizardEvent& evt )
 {
 	if( ((sptr)evt.GetPage() == (sptr)&m_page_bios) )
-		m_panel_BiosSel.OnShown();
+	    m_panel_BiosSel.OnShown();
```
Comment (lightningterror) [inline pcsx2/gui/Dialogs/FirstTimeWizard.cpp:231]: Spaces.

**Comment by RedDevilus** [comment]
Formatting should be fine now

File: common/include/Utilities/wxGuiTools.h [hunk 4]
```diff
@@ -257,7 +257,6 @@ void operator+=(wxWindow &target, WinType &src)
         return;
     *target.GetSizer() += src;
 }
-
```
Comment (lightningterror) [inline common/include/Utilities/wxGuiTools.h:260]: This too, useless.

Comment (RedDevilus) [reply common/include/Utilities/wxGuiTools.h]: Fixed

File: pcsx2/gui/Dialogs/SysConfigDialog.cpp [hunk 5]
```diff
@@ -192,7 +192,7 @@ void Dialogs::SysConfigDialog::Cancel()
 }
 
 Dialogs::SysConfigDialog::SysConfigDialog(wxWindow* parent)
-	: BaseConfigurationDialog( parent, AddAppName(_("Emulation Settings - %s")), 580 )
+	: BaseConfigurationDialog( parent, AddAppName(_("General Settings - %s")), 600 )
```
Comment (lightningterror) [inline pcsx2/gui/Dialogs/SysConfigDialog.cpp:195]: Hm, why the 600 change?

Comment (RedDevilus) [reply pcsx2/gui/Dialogs/SysConfigDialog.cpp:195]: Testing but forgot

File: pcsx2/gui/Panels/PathsPanel.cpp [hunk 6]
```diff
@@ -79,3 +74,13 @@ Panels::StandardPathsPanel::StandardPathsPanel( wxWindow* parent )
...
 
+void Panels::StandardPathsPanel::Apply()
+{
+
+}
+
+void Panels::StandardPathsPanel::AppStatusEvent_OnSettingsApplied()
+{
+
+}
```
Comment (lightningterror) [inline pcsx2/gui/Panels/PathsPanel.cpp:85]: What purpose do these functions serve?

Comment (RedDevilus) [reply pcsx2/gui/Panels/PathsPanel.cpp:85]: This fixes the warning in first time wizard and the function to block users to not press next until they provide a valid bios.

Comment (RedDevilus) [reply pcsx2/gui/Panels/PathsPanel.cpp:85]: It relies on the fuction  onselbios, a cursor or atleast an update to the screen where you pick a bios.

So it handles this:
image

Whoever created this made it hardcoded and obfuscated.

Comment (GovanifY) [reply pcsx2/gui/Panels/PathsPanel.cpp:85]: Yep, I can confirm those are necessary, they are assumed to be defined by inheritance. Other windows have them dummied out too; blame wx/air

**Review by lightningterror — APPROVED** [review]
Looks good, just need one last round of testing.

File: pcsx2/gui/Dialogs/SysConfigDialog.cpp [hunk 7]
```diff
@@ -202,6 +202,8 @@ Dialogs::SysConfigDialog::SysConfigDialog(wxWindow* parent)
 
 	//NOTE: all pages which are added to SysConfigDialog must be of class BaseApplicableConfigPanel or derived.
 	//		see comment inside UpdateGuiForPreset implementation for more info.
+	AddPage<BiosSelectorPanel>		( pxL("BIOS"),			cfgid.Cpu );
+	AddPage<StandardPathsPanel>		( pxL("Folders"),		cfgid.Paths );
```
Comment (tadanokojin) [inline pcsx2/gui/Dialogs/SysConfigDialog.cpp:206]: formatting

Comment (RedDevilus) [reply pcsx2/gui/Dialogs/SysConfigDialog.cpp:206]: image

How is it displaying it like that?

Comment (tadanokojin) [reply pcsx2/gui/Dialogs/SysConfigDialog.cpp:206]: tabs sucking at aligning things maybe, it's not really a huge issue.

**Review by tadanokojin — APPROVED** [review]
LGTM. @GovanifY can you verified your requested change has been made?

**Review by GovanifY — APPROVED** [review]
Yep, all good!

**Pull request merged**